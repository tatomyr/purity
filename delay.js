export const delay = (t = 0) => ({
    then: (resolve) => {
        const timeout = setTimeout(() => resolve(timeout), t);
    },
});
