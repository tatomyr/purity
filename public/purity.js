const clearFalsy = (x) => x === undefined || x === null || x === false ? '' : x;
const joinIfArray = (x) => Array.isArray(x) ? x.join('') : x;
const PURITY_KEYWORD = 'purity';
const DATA_PURITY_FLAG = `data-${PURITY_KEYWORD}_flag`;
export const init = (initialState) => {
    let state = initialState;
    const parseHTML = (html) => {
        const virtualDocument = new DOMParser().parseFromString(html, 'text/html');
        const nodesMap = new Map();
        for (const node of virtualDocument.querySelectorAll('[id]')) {
            let shallow = node.cloneNode(true);
            for (let innerNode of shallow.querySelectorAll('[id]')) {
                innerNode.outerHTML = `<!-- ${innerNode.tagName}#${innerNode.id} -->`;
            }
            console.log('shallow--->', shallow.outerHTML);
            for (let innerNode of shallow.querySelectorAll(`[${DATA_PURITY_FLAG}]`)) {
                console.log('         |-->', innerNode.outerHTML);
                for (const key in innerNode.dataset) {
                    if (key.startsWith(PURITY_KEYWORD)) {
                        innerNode.removeAttribute(`data-${key}`);
                    }
                }
            }
            nodesMap.set(node.id, { node, shallow });
        }
        return nodesMap;
    };
    let rootComponent;
    let domNodesMap;
    function mount(f) {
        rootComponent = f;
        domNodesMap = parseHTML(rootComponent());
        const rootId = domNodesMap.keys().next().value;
        const root = document.getElementById(rootId);
        const rootNode = domNodesMap.get(rootId)?.node;
        if (root && rootNode) {
            root.replaceWith(rootNode);
        }
        else {
            throw new Error(`Root DOM element's id does not correspond to the defined application root id (${rootId}).`);
        }
    }
    function updateAttributes(element, newNode) {
        for (const { name } of element.attributes) {
            element.removeAttribute(name);
        }
        for (const { name, value } of newNode.node.attributes) {
            element.setAttribute(name, value);
        }
    }
    function rerender() {
        const newNodesMap = parseHTML(rootComponent());
        for (const [id, domNode] of domNodesMap) {
            const newNode = newNodesMap.get(id);
            if (newNode && domNode.shallow.outerHTML !== newNode.shallow.outerHTML) {
                const elementById = document.getElementById(id);
                if (elementById) {
                    updateAttributes(elementById, newNode);
                    if (domNode.shallow.innerHTML !== newNode.shallow.innerHTML) {
                        elementById.innerHTML = newNode.node.innerHTML;
                        console.log(`↻ #${id}`);
                    }
                    else {
                        console.log(`± #${id}`);
                    }
                }
                else {
                    throw new Error(`There is no element in DOM with id ${id}.`);
                }
            }
        }
        domNodesMap = newNodesMap;
    }
    return {
        mount,
        rerender,
        getState: () => state,
        setState(callback) {
            Object.assign(state, callback(state));
            rerender();
        },
    };
};
const ARGS_RE = /__\[(\d+)\]__/gm;
const BOUND_EVENTS_RE = /::(\w+)\s*=\s*__\[(\d+)\]__/gm;
const applyPurityKey = (() => {
    let purityKey = 0;
    let timeout;
    return () => {
        if (timeout) {
            clearTimeout(timeout);
        }
        timeout = setTimeout(() => {
            purityKey = 0;
        });
        return purityKey++;
    };
})();
export const render = ([first, ...strings], ...args) => {
    const precomputedString = strings.reduce(($, item, i) => `${$}__[${i}]__${item}`, first);
    const bindEventHandlers = (_, event, index) => {
        const dataName = `data-${PURITY_KEYWORD}_${event}_${applyPurityKey()}`;
        setTimeout(() => {
            let element = document.querySelector(`[${dataName}]`);
            if (element) {
                ;
                element[`on${event}`] = args[index];
                element.removeAttribute(dataName);
            }
        });
        return `${dataName} ${DATA_PURITY_FLAG}`;
    };
    const processArgs = (_, index) => joinIfArray(clearFalsy(args[+index]));
    const stringToRender = precomputedString
        .replace(BOUND_EVENTS_RE, bindEventHandlers)
        .replace(ARGS_RE, processArgs)
        .trim()
        .replace(/\n\s*</g, '<')
        .replace(/>\n\s*/g, '>');
    return stringToRender;
};
