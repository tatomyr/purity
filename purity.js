// Constants
const PURITY_KEYWORD = "purity";
const DATA_PURITY_FLAG = `data-${PURITY_KEYWORD}_flag`;
/**
 * App factory that should be invoked once to create an application state
 */
export const init = (initialState) => {
    const state = initialState;
    /**
     * Parses an html string and returns so called 'nodeMap' which represents the virtual DOM
     */
    const buildNodesMap = (html) => {
        const virtualDocument = new DOMParser().parseFromString(html, "text/html");
        const nodesMap = new Map();
        for (const node of virtualDocument.querySelectorAll("[id]")) {
            const shallow = node.cloneNode(true);
            for (const innerNode of shallow.querySelectorAll("[id]")) {
                innerNode.outerHTML = `<!-- ${innerNode.tagName}#${innerNode.id} -->`;
            }
            // Removing the `data-purity_*` attributes attached in render() function
            // TODO: try to avoid the situation when we have to remove something added in another module.
            for (const innerNode of shallow.querySelectorAll(`[${DATA_PURITY_FLAG}]`)) {
                for (const key in innerNode.dataset) {
                    if (key.startsWith(PURITY_KEYWORD)) {
                        innerNode.removeAttribute(`data-${key}`);
                    }
                }
            }
            nodesMap.set(node.id, { node, shallow });
        }
        return nodesMap;
    };
    let rootComponent;
    let domNodesMap;
    /**
     * Mounts an App to DOM
     */
    function mount(f) {
        // Setting up rootComponent
        rootComponent = () => buildNodesMap(f());
        domNodesMap = rootComponent();
        // Top-level component should always have an id equal to a root element's id
        const rootId = domNodesMap.keys().next().value;
        const root = document.getElementById(rootId);
        const rootNode = domNodesMap.get(rootId)?.node;
        if (root && rootNode) {
            root.replaceWith(rootNode);
        }
        else {
            throw new Error(`Root DOM element's id does not correspond to the defined application root id "${rootId}".`);
        }
    }
    /**
     * Updates element's attributes from current state to one specified in newNode
     */
    function updateAttributes(element, newNode) {
        for (const { name } of element.attributes) {
            element.removeAttribute(name);
        }
        for (const { name, value } of newNode.node.attributes) {
            element.setAttribute(name, value);
        }
    }
    /**
     * Forces html re-rendering with the current state
     */
    function rerender() {
        const newNodesMap = rootComponent();
        console.warn("ðŸŒ€");
        for (const [id, domNode] of domNodesMap) {
            const newNode = newNodesMap.get(id);
            // Since we depend on the shallow comparison, we must only care about updating changed nodes.
            if (newNode && domNode.shallow.outerHTML !== newNode.shallow.outerHTML) {
                const elementById = document.getElementById(id);
                if (elementById) {
                    updateAttributes(elementById, newNode);
                    if (domNode.shallow.innerHTML !== newNode.shallow.innerHTML) {
                        elementById.innerHTML = newNode.node.innerHTML;
                        console.warn(`\tâ†» #${id}`);
                    }
                    else {
                        console.warn(`\tÂ± #${id}`);
                    }
                }
                else {
                    throw new Error(`There is no element in the DOM with id "${id}".`);
                }
            }
        }
        domNodesMap = newNodesMap;
    }
    return {
        mount,
        rerender,
        getState: () => state,
        setState(callback) {
            Object.assign(state, callback(state));
            rerender();
        },
    };
};
// Patterns
const ARGS_RE = /__\[(\d+)\]__/gm;
const BOUND_EVENTS_RE = /::(\w+)\s*=\s*__\[(\d+)\]__/gm;
// Helpers
export const isTruthy = (x) => x !== undefined && x !== null && x !== false;
const clearFalsy = (x) => isTruthy(x) ? x : "";
const joinIfArray = (x) => Array.isArray(x) ? x.join("") : x;
/**
 * Increments the Purity Key and resets it after all sync operations completed
 */
const applyPurityKey = (() => {
    let purityKey = 0;
    let timeout;
    return () => {
        if (timeout) {
            clearTimeout(timeout);
        }
        timeout = setTimeout(() => {
            purityKey = 0;
        });
        return purityKey++;
    };
})();
/**
 * Tagged template to compute the html string from a string literal
 */
export const render = ([first, ...strings], ...args) => {
    const precomputedString = strings.reduce(($, item, i) => `${$}__[${i}]__${item}`, first);
    const bindEventHandlers = (_, event, index) => {
        const dataName = `data-${PURITY_KEYWORD}_${event}_${applyPurityKey()}`;
        setTimeout(() => {
            // Asynchronously bind event handlers after rendering everything to DOM
            const element = document.querySelector(`[${dataName}]`);
            const prop = args[index];
            if (element && typeof prop === "function") {
                element[`on${event}`] = prop;
                // Remove residuals (needed for consistency)
                element.removeAttribute(dataName);
            }
        });
        return `${dataName} ${DATA_PURITY_FLAG}`;
    };
    const processArgs = (_, index) => joinIfArray(clearFalsy(args[+index]));
    const stringToRender = precomputedString
        .replace(BOUND_EVENTS_RE, bindEventHandlers)
        .replace(ARGS_RE, processArgs)
        .trim()
        .replace(/\n\s*</g, "<")
        .replace(/>\n\s*/g, ">");
    return stringToRender;
};
