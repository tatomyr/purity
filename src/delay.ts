export const delay = (t: number): Promise<void> =>
  ({
    then: (resolve: typeof Promise.resolve) => {
      setTimeout(resolve, t)
    },
  } as Promise<void>)
