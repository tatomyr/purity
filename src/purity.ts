// Types
type Allowed = string | number
type Rejected = undefined | null | false
type Simple = Allowed | Rejected
type Argument = Simple | string[]
type Verified = Allowed | string[]
export type Component = <P>(props?: P, ...rest: any[]) => string // FIXME: it doesn't allow calling without an argument
type VirtualNodes = {
  node: HTMLElement
  shallow: HTMLElement
}
type DomNodesMap = Map<string, VirtualNodes> // FIXME: Element/HTMLElement - pick one
export type App<State> = {
  mount: (f: Component) => void
  rerender: () => void
  getState: () => State
  setState(callback: (state: State) => Partial<State>): void
}

// Constants
const PURITY_KEYWORD = 'purity'
const DATA_PURITY_FLAG = `data-${PURITY_KEYWORD}_flag`

/**
 * App factory that should be invoked once to create a single store with reactive state
 */
export const init = <State extends Record<string, unknown>>(
  initialState: State
): App<State> => {
  // eslint-disable-next-line prefer-const
  let state = initialState

  /**
   * Parses html string and returns so called 'nodeMap' which represents virtual DOM
   */
  const buildNodesMap = (html: string): DomNodesMap => {
    const virtualDocument = new DOMParser().parseFromString(html, 'text/html')
    const nodesMap: DomNodesMap = new Map()
    for (const node of virtualDocument.querySelectorAll('[id]')) {
      // eslint-disable-next-line prefer-const
      let shallow = (node as HTMLElement).cloneNode(true) as HTMLElement // FIXME: null?
      // eslint-disable-next-line prefer-const
      for (let innerNode of shallow.querySelectorAll('[id]')) {
        innerNode.outerHTML = `<!-- ${innerNode.tagName}#${innerNode.id} -->`
      }
      // Removing the `data-purity_*` attributes attached in render() function
      // TODO: try to avoid the situation when we have to remove something added in another module.
      // eslint-disable-next-line prefer-const
      for (let innerNode of shallow.querySelectorAll(`[${DATA_PURITY_FLAG}]`)) {
        for (const key in (innerNode as HTMLElement).dataset) {
          if (key.startsWith(PURITY_KEYWORD)) {
            innerNode.removeAttribute(`data-${key}`)
          }
        }
      }
      nodesMap.set(node.id, {node, shallow} as VirtualNodes)
    }
    return nodesMap
  }

  let rootComponent: () => DomNodesMap
  let domNodesMap: DomNodesMap
  /**
   * Mounts an App to DOM
   */
  function mount(f: Component) {
    // Setting up rootComponent
    rootComponent = () => buildNodesMap(f())
    domNodesMap = rootComponent()
    // Top-level component should always have an id equal to a root element's id
    const rootId: string = domNodesMap.keys().next().value
    const root = document.getElementById(rootId)
    const rootNode = domNodesMap.get(rootId)?.node
    if (root && rootNode) {
      root.replaceWith(rootNode)
    } else {
      throw new Error(
        `Root DOM element's id does not correspond to the defined application root id "${rootId}".`
      )
    }
  }

  /**
   * Updates element's attributes from current state to one specified in newNode
   */
  function updateAttributes(element: HTMLElement, newNode: VirtualNodes) {
    for (const {name} of element.attributes) {
      element.removeAttribute(name)
    }
    for (const {name, value} of newNode.node.attributes) {
      element.setAttribute(name, value)
    }
  }

  /**
   * Forces html re-rendering with the current state
   */
  function rerender() {
    const newNodesMap = rootComponent()
    for (const [id, domNode] of domNodesMap) {
      const newNode = newNodesMap.get(id)
      // Since we depend on the shallow comparison, we must only care about updating changed nodes.
      if (newNode && domNode.shallow.outerHTML !== newNode.shallow.outerHTML) {
        const elementById = document.getElementById(id)
        if (elementById) {
          updateAttributes(elementById, newNode)
          if (domNode.shallow.innerHTML !== newNode.shallow.innerHTML) {
            elementById.innerHTML = newNode.node.innerHTML
            console.log(`↻ #${id}`)
          } else {
            console.log(`± #${id}`)
          }
        } else {
          throw new Error(`There is no element in DOM with id "${id}".`)
        }
      }
    }
    domNodesMap = newNodesMap
  }

  return {
    mount,
    rerender,
    getState: () => state,
    setState(callback: (state: State) => Partial<State>) {
      Object.assign(state, callback(state))
      rerender()
    },
  }
}

// Patterns
const ARGS_RE = /__\[(\d+)\]__/gm
const BOUND_EVENTS_RE = /::(\w+)\s*=\s*__\[(\d+)\]__/gm

// Helpers
const clearFalsy = <T extends Verified>(x: T | Rejected): T | '' =>
  x === undefined || x === null || x === false ? '' : x

const joinIfArray = (x: Verified): Allowed =>
  Array.isArray(x) ? x.join('') : x

/**
 * Increases the Purity Key and resets it after all sync operations completed
 */
const applyPurityKey = (() => {
  let purityKey = 0
  let timeout: number
  return () => {
    if (timeout) {
      clearTimeout(timeout)
    }
    timeout = setTimeout(() => {
      purityKey = 0
    })
    return purityKey++
  }
})()

/**
 * Tagged template to compute the html string from a string literal
 */
export const render = (
  [first, ...strings]: TemplateStringsArray,
  ...args: Array<Argument | EventHandlerNonNull>
): string => {
  const precomputedString: string = strings.reduce(
    ($, item, i) => `${$}__[${i}]__${item}`,
    first
  )

  const bindEventHandlers = (
    _: any,
    event: keyof HTMLElementEventMap,
    index: number
  ) => {
    const dataName = `data-${PURITY_KEYWORD}_${event}_${applyPurityKey()}`
    setTimeout(() => {
      // Asynchronously bind event handlers after rendering everything to DOM
      // eslint-disable-next-line prefer-const
      let element: HTMLElement | null = document.querySelector(`[${dataName}]`)
      if (element) {
        ;(element as {[key: string]: any})[`on${event}`] = args[
          index
        ] as EventHandlerNonNull
        // Remove residuals
        element.removeAttribute(dataName)
      }
    })
    return `${dataName} ${DATA_PURITY_FLAG}`
  }

  const processArgs = (_: any, index: number): string =>
    joinIfArray(clearFalsy(args[+index] as Argument)) as string

  const stringToRender = precomputedString
    .replace(BOUND_EVENTS_RE, bindEventHandlers)
    .replace(ARGS_RE, processArgs)
    .trim()
    .replace(/\n\s*</g, '<')
    .replace(/>\n\s*/g, '>')

  return stringToRender
}
