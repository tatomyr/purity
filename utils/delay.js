export const delay = t => ({ then: resolve => setTimeout(resolve, t) })
