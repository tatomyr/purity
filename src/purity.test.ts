// const {init, render} = require('./__purity__.js')
import {App, Component, init, render} from './purity'
import {delay} from './delay'

export type AnyObject = {[key: string]: any}

// const delay = t => ({then: resolve => setTimeout(resolve, t)})

// const SimpleComponent = () => render`<div id="root">SOMETHING</div>`
// const ComplexComponent = ({something}) =>
//   render`
//     <div id="root">
//       <h1 id="title">TITLE</h1>
//       <p id="something">${something}</p>
//     </div>
//   `
// const DymamicComponent = ({something}) =>
//   render`
//     <div id="root">
//       <h1 id="title">TITLE</h1>
//       <p id="something" title="${something}">${something}</p>
//     </div>
//   `
// const ConditionalComponent = ({something}) =>
//   render`
//     <div id="root">
//       ${something && render`<p id="something">${something}</p>`}
//     </div>
//   `

describe('purity', () => {
  let app: App<AnyObject>, defaultState: AnyObject
  beforeEach(() => {
    document.body.innerHTML = '<div id="root"></div>'
  })
  afterEach(() => {
    document.body.innerHTML = ''
    app = undefined
  })

  describe('counter app', () => {
    let CounterApp: Component
    beforeEach(() => {
      defaultState = {title: 'COUNTER', counter: 0}
      app = init(defaultState)
      CounterApp = () => render`
        <div id="root">
          <h1>${app.getState().title}</h1>
          <span id="count">${app.getState().count}</span>
        </div>
      `
    })
    it('should match default state after created', () => {
      expect(app.getState()).toEqual(defaultState)
    })
    it('should mount a component depending on the default state', () => {
      app.mount(CounterApp)
      expect(document.body.innerHTML).toEqual(CounterApp())
    })
    it('should change state after updates depending on previous state', () => {
      app.mount(CounterApp)
      app.setState(state => ({counter: state.counter + 1}))
      expect(app.getState()).toEqual({title: 'COUNTER', counter: 1})
    })
  })

  it('should bind an event', async () => {
    let eventHandler = jest.fn()
    const ClickableComponent = () => render`
      <button id="root" ::click=${eventHandler}>Click me</button>
    `
    // app = init({})
    app.mount(ClickableComponent)
    // Awaiting for the eventHandler to be set in setTimeout
    await delay(0)
    expect(document.body.innerHTML).toMatchSnapshot()
    // @ts-ignore
    document.querySelector('button#root').click()
    expect(eventHandler).toHaveBeenCalled()
  })
  // it('should bind multiple events', async () => {
  //   let clickHandler = jest.fn()
  //   let blurHandler = jest.fn()
  //   const ClickableComponent = () => render`
  //     <input type="text" id="root" ::click=${clickHandler} ::blur=${blurHandler} />
  //   `
  //   app.mount(ClickableComponent)
  //   // Awaiting for the eventHandler to be set in setTimeout
  //   await delay(0)
  //   expect(document.body.innerHTML).toMatchSnapshot()
  //   document.querySelector('input#root').click()
  //   expect(clickHandler).toHaveBeenCalled()
  //   document.querySelector('input#root').focus()
  //   document.querySelector('input#root').blur()
  //   expect(blurHandler).toHaveBeenCalled()
  // })

  // it('should handle changes for a component with a static wrapper', () => {
  //   app.mount(app.connect(ComplexComponent))
  //   app.dispatch({type: 'CHANGE_SOMETHING', something: 'new thing'})
  //   expect(document.body.innerHTML).toEqual(
  //     ComplexComponent({something: 'new thing'})
  //   )
  // })
  // it('should handle changes for a component with a dynamic wrapper', () => {
  //   app.mount(app.connect(DymamicComponent))
  //   app.dispatch({type: 'CHANGE_SOMETHING', something: 'new thing'})
  //   expect(document.body.innerHTML).toEqual(
  //     DymamicComponent({something: 'new thing'})
  //   )
  // })
  // it('should handle a conditional component', () => {
  //   app.mount(app.connect(ConditionalComponent))
  //   app.dispatch({type: 'CHANGE_SOMETHING', something: ''})
  //   expect(document.body.innerHTML).toEqual(
  //     ConditionalComponent({something: ''})
  //   )
  //   app.dispatch({type: 'CHANGE_SOMETHING', something: 'new thing'})
  //   expect(document.body.innerHTML).toEqual(
  //     ConditionalComponent({something: 'new thing'})
  //   )
  // })

  // it('should deal with a compound component', () => {
  //   const InnerComponent = app.connect(
  //     ({something, style}) => render`
  //       <h1 id="title" style="${style}">TITLE</h1>
  //       <p id="something">${something}</p>
  //     `
  //   )
  //   const CompundComponent = () => render`
  //     <div id="root">
  //       ${InnerComponent({style: 'color: red;'})}
  //     </div>
  //   `
  //   app.mount(CompundComponent)
  //   expect(document.body.innerHTML).toMatchSnapshot()
  // })
  // it('should bind an event', async () => {
  //   let eventHandler = jest.fn()
  //   const ClickableComponent = () => render`
  //     <button id="root" ::click=${eventHandler}>Click me</button>
  //   `
  //   app.mount(ClickableComponent)
  //   // Awaiting for the eventHandler to be set in setTimeout
  //   await delay(0)
  //   expect(document.body.innerHTML).toMatchSnapshot()
  //   document.querySelector('button#root').click()
  //   expect(eventHandler).toHaveBeenCalled()
  // })
  // it('should bind multiple events', async () => {
  //   let clickHandler = jest.fn()
  //   let blurHandler = jest.fn()
  //   const ClickableComponent = () => render`
  //     <input type="text" id="root" ::click=${clickHandler} ::blur=${blurHandler} />
  //   `
  //   app.mount(ClickableComponent)
  //   // Awaiting for the eventHandler to be set in setTimeout
  //   await delay(0)
  //   expect(document.body.innerHTML).toMatchSnapshot()
  //   document.querySelector('input#root').click()
  //   expect(clickHandler).toHaveBeenCalled()
  //   document.querySelector('input#root').focus()
  //   document.querySelector('input#root').blur()
  //   expect(blurHandler).toHaveBeenCalled()
  // })
  // it(`
  //   should not change innerHTML when only attributes have changed in the wrapper tag
  //   (input's value should remain the same)
  // `, async () => {
  //   const handleClick = e => {
  //     app.dispatch({
  //       type: 'CHANGE_SOMETHING',
  //       something: document.querySelector('#color').value,
  //     })
  //   }

  //   const Component = app.connect(
  //     ({something}) => render`
  //       <div id="root">
  //         <input id="color" style="color: ${something};" />
  //         <button
  //           ::click=${handleClick}
  //         >
  //           Apply color
  //         </button>
  //       </div>
  //     `
  //   )
  //   app.mount(Component)
  //   await delay(0)
  //   document.querySelector('#color').value = 'red'
  //   await delay(0)
  //   document.querySelector('button').onclick()
  //   await delay(0)
  //   expect(document.querySelector('#color').value).toEqual('red')
  //   expect(document.querySelector('#color').style.color).toEqual('red')
  //   expect(document.body.innerHTML).toMatchSnapshot()
  // })
  // it('should handle conditional rendering & process arrays', () => {
  //   const Component = ({maybeArr}) => render`
  //     <div id="root">
  //       <ul>
  //         ${!!maybeArr && maybeArr.map(item => render`<li>${item}</li>`)}
  //       </ul>
  //     </div>
  //   `
  //   app.mount(() => Component({}))
  //   expect(document.body.innerHTML).toMatchSnapshot()
  //   app.mount(() => Component({maybeArr: ['üçé', 'üçå', 'üç∞']}))
  //   expect(document.body.innerHTML).toMatchSnapshot()
  // })
})
