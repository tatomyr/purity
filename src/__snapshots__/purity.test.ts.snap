// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`purity counter app should deal with nested components 1`] = `"<div id=\\"root\\"><h1>COUNTER</h1><span id=\\"count\\"></span><button id=\\"button\\">Increment</button></div>"`;

exports[`purity should bind an event 1`] = `"<button id=\\"root\\" data-purity_flag=\\"\\">Click me</button>"`;

exports[`purity should bind multiple events 1`] = `"<input type=\\"text\\" id=\\"root\\" data-purity_flag=\\"\\">"`;

exports[`purity should handle conditional rendering & process arrays 1`] = `"<div id=\\"root\\"><ul></ul></div>"`;

exports[`purity should handle conditional rendering & process arrays 2`] = `"<div id=\\"root\\"><ul><li>🍎</li><li>🍌</li><li>🍰</li></ul></div>"`;
